core/shared/src/main/scala/zio/State.scala

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2021 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio
+
+/**
+ * `State[S]` models a value of type `S` that can be read from and written to
+ * during the execution of an effect. The idiomatic way to work with `State`
+ * is as part of the environment using operators defined on `ZIO`. For
+ * example:
+ *
+ * {{{
+ * final case class MyState(counter: Int)
+ *
+ * for {
+ *   _     <- ZIO.updateState[MyState](state => state.copy(counter = state.counter + 1))
+ *   count <- ZIO.getStateWith[MyState](_.counter)
+ * } yield count
+ * }}}
+ *
+ * Because `State` is typically used as part of the environment, it is
+ * recommended to define your own state type `S` such as `MyState` above
+ * rather than using a type such as `Int` to avoid the risk of ambiguity.
+ *
+ * To run an effect that depends on some state, create the initial state with
+ * the `make` constructor and then use `toLayer` to convert it into a layer
+ * that you can provide along with your application's other dependencies.
+ */
+sealed trait State[S] {
```

Comment:
This could be generalized to `ZState[E1, E2, S1, S2]` and then `type State[S] = ZState[Nothing, Nothing, S, S]`. I'm not sure how useful this would be and I think for ergonomics it would require another set of operators on `ZIO` (e.g. `getZState`, `setZState`) to avoid having to specify additional type parameters in the common case.

Reply:
I could see this proliferating very fast in the `R` parameter, it would be great if either:
- `State` was renamed to `ST` ([homage](https://wiki.haskell.org/Monad/ST)), or
- the state was composable, though you'd have to think of how this would interfere with the semantics of `FiberRef`

Reply:
The other consideration is that it would place additional strain on overloading (the highly polymorphic ZState).

Still if there's a good argument, now is the time to consider it.

Neutral on rename to `ST` but I do think it needs to be renamed. 👍

Reply:
Renamed to `ZState`. I think it is nice to have names that are slightly more descriptive, especially since it is easy to use type aliases for your state type or types.